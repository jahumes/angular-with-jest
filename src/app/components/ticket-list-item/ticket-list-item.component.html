<div class="d-flex">
    <div class="flex-grow-1">
        <a [routerLink]="'/ticket/' + ticket.id">{{ ticket.description }}</a>
    </div>
    <div class="mr-3" [formGroup]="assigneeForm">
        <select name="userId" class="form-control" formControlName="assigneeId">
            <option [value]="null">--Assign User--</option>
            <option *ngFor="let user of users" [value]="user.id">{{
                user.name
            }}</option>
        </select>
    </div>
    <div>
        <button
            *ngIf="ticket.completed === false"
            class="btn btn-secondary btn-sm"
            (click)="completeTicket($event)"
        >
            Complete
        </button>
        <button
            *ngIf="ticket.completed === true"
            class="btn btn-secondary btn-sm"
            (click)="reopenTicket($event)"
        >
            Reopen
        </button>
    </div>
</div>
