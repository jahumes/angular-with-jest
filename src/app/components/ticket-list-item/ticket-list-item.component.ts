import { Component, Input, OnInit } from "@angular/core";
import { TicketFacadeService } from "../../services/ticket-facade.service";
import { Ticket, User } from "../../backend.service";
import { FormControl, FormGroup } from "@angular/forms";

@Component({
    selector: "app-ticket-list-item",
    templateUrl: "./ticket-list-item.component.html",
    styleUrls: ["./ticket-list-item.component.css"]
})
export class TicketListItemComponent implements OnInit {
    @Input() ticket: Ticket;
    @Input() users: User[];

    assigneeForm = new FormGroup({
        assigneeId: new FormControl(null)
    });

    constructor(private _ticketFacade: TicketFacadeService) {}

    ngOnInit() {
        this.assigneeForm.setValue({
            assigneeId: this.ticket.assigneeId
        });

        this.assigneeForm
            .get("assigneeId")
            .valueChanges.subscribe(assigneeId =>
                this._ticketFacade.assignTicket(this.ticket.id, +assigneeId)
            );
    }

    completeTicket($event: any) {
        $event.preventDefault();
        this._ticketFacade.completeTicket(this.ticket.id);
    }

    reopenTicket($event: any) {
        $event.preventDefault();
        this._ticketFacade.reopenTicket(this.ticket.id);
    }
}
