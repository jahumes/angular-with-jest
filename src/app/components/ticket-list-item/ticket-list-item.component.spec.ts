import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { TicketListItemComponent } from "./ticket-list-item.component";
import { RouterTestingModule } from "@angular/router/testing";
import { TicketFacadeService } from "../../services/ticket-facade.service";
import { ReactiveFormsModule } from "@angular/forms";

describe("TicketListItemComponent", () => {
    let component: TicketListItemComponent;
    let fixture: ComponentFixture<TicketListItemComponent>;
    let ticketFacade: any;

    beforeEach(async(() => {
        ticketFacade = {
            completeTicket: jest.fn()
        };
        TestBed.configureTestingModule({
            declarations: [TicketListItemComponent],
            imports: [RouterTestingModule, ReactiveFormsModule],
            providers: [
                {
                    provide: TicketFacadeService,
                    useValue: ticketFacade
                }
            ]
        }).compileComponents();
    }));

    beforeEach(() => {
        fixture = TestBed.createComponent(TicketListItemComponent);
        component = fixture.componentInstance;
        component.ticket = {
            id: 1,
            description: "description",
            assigneeId: null,
            completed: false
        };
        fixture.detectChanges();
    });

    it("should create", () => {
        expect(component).toBeTruthy();
    });

    it("should have a Complete button if the ticket is open and Reopen if the ticket is complete", () => {
        expect(fixture).toMatchSnapshot();

        component.ticket.completed = true;

        fixture.detectChanges();

        expect(fixture).toMatchSnapshot();
    });
});
