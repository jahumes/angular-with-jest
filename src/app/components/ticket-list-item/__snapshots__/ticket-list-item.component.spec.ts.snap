// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketListItemComponent should have a Complete button if the ticket is open and Reopen if the ticket is complete 1`] = `
<app-ticket-list-item
  _ticketFacade={[Function Object]}
  assigneeForm={[Function FormGroup]}
  ticket={[Function Object]}
>
  <div
    class="d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <a
        href="/ticket/1"
        ng-reflect-router-link="/ticket/1"
      >
        description
      </a>
    </div>
    <div
      class="mr-3 ng-untouched ng-pristine ng-valid"
      ng-reflect-form="[object Object]"
    >
      <select
        class="form-control ng-untouched ng-pristine ng-valid"
        formcontrolname="assigneeId"
        name="userId"
        ng-reflect-name="assigneeId"
      >
        <option
          value="null"
        >
          --Assign User--
        </option>
        
      </select>
    </div>
    <div>
      
      <button
        class="btn btn-secondary btn-sm"
      >
         Complete 
      </button>
      
    </div>
  </div>
</app-ticket-list-item>
`;

exports[`TicketListItemComponent should have a Complete button if the ticket is open and Reopen if the ticket is complete 2`] = `
<app-ticket-list-item
  _ticketFacade={[Function Object]}
  assigneeForm={[Function FormGroup]}
  ticket={[Function Object]}
>
  <div
    class="d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <a
        href="/ticket/1"
        ng-reflect-router-link="/ticket/1"
      >
        description
      </a>
    </div>
    <div
      class="mr-3 ng-untouched ng-pristine ng-valid"
      ng-reflect-form="[object Object]"
    >
      <select
        class="form-control ng-untouched ng-pristine ng-valid"
        formcontrolname="assigneeId"
        name="userId"
        ng-reflect-name="assigneeId"
      >
        <option
          value="null"
        >
          --Assign User--
        </option>
        
      </select>
    </div>
    <div>
      
      
      <button
        class="btn btn-secondary btn-sm"
      >
         Reopen 
      </button>
    </div>
  </div>
</app-ticket-list-item>
`;
